[Merged by Bors] - feat(Data/Matrix/Cartan): add explicit forms for B, C, D

JohnnyTeutonic opened the pull request:
Add explicit matrix representations for small Cartan matrices:
- B_two: B Cartan matrix
- C_two: C Cartan matrix
- D_four: D Cartan matrix (minimal rank for type D)


---
<!-- Your PR title will become the first line of the commit message.

In this box, the text above the `---` (if not empty) will be appended
to the commit message, and can be used to give additional context or
details. Please leave a blank newline before the `---`, otherwise GitHub
will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

JohnnyTeutonic committed:
feat(Data/Matrix/Cartan): add explicit forms for B, C, D

Add explicit matrix representations for small Cartan matrices:
- B_two: B Cartan matrix
- C_two: C Cartan matrix
- D_four: D Cartan matrix (minimal rank for type D)

JohnnyTeutonic committed:
Merge branch 'master' into cartan-matrix-properties

JohnnyTeutonic committed:
Merge branch 'master' into cartan-matrix-properties

JohnnyTeutonic committed:
Merge branch 'master' into cartan-matrix-properties

JohnnyTeutonic committed:
Merge branch 'master' into cartan-matrix-properties

ocfnash commented on Mathlib/Data/Matrix/Cartan.lean:
```diff
@@ -175,6 +175,19 @@ theorem A_two : A 2 = !![2, -1; -1, 2] := by
+
+/-- The Cartan matrix D in explicit form. -/
+theorem D_four : D 4 = !![2, -1, 0, 0; -1, 2, -1, -1; 0, -1, 2, 0; 0, -1, 0, 2] := by
+  ext i j; fin_cases i <;> fin_cases j <;> simp [D, Matrix.of_apply]
```
Come to think of it, I think we should have `D 2 = !![2, 0; 0, 2]` rather than `D 2 = A 2`.

JohnnyTeutonic replied:
Thanks for the detailed suggestions.

I'll apply:
- Aligned matrix formatting with `decide`
- Drop the docstrings
- Add `B_one`, `C_one`, `D_one`
- `D_two = !![2, 0; 0, 2]` (per your correction - disconnected diagram)

For the `D` definition bug affecting `D_three` - should I attempt that fix in this PR, or would you prefer a separate PR for that?

ocfnash replied:
Glancing again now, it occurs to be that `D 3` need not actually equal `A 3` and perhaps it might be more natural / convenient for `D 3` to be `[2, -1, -1; -1,  2,  0; -1,  0,  2]`. I'm happy as long as the two matrices are equivalent up to reindexing (i.e., permutation of the nodes in the Dynkin diagram). So we just need:
```lean
example : (D 3).reindex c[0, 1] c[0, 1] = A 3 := by decide
```
(You'll need to add `public import Mathlib.GroupTheory.Perm.Cycle.Concrete` for the notation here.)

JohnnyTeutonic committed:
refactor(Data/Matrix/Cartan): improve formatting and add coincidence lemmas

- Use decide tactic and human-friendly matrix formatting
- Add B_one, C_one coincidence lemmas
- Add D_three explicit form and reindex equivalence with A_three
- Add import for Perm.Cycle.Concrete for reindex notation

ocfnash commented:
@JohnnyTeutonic This looks good except for the fix for `D 2`. Are you still planning on getting to that?

JohnnyTeutonic committed:
fix(Data/Matrix/Cartan): fix D definition for small n and add D_one, D_two

- Fix D definition to handle n  2 correctly (was broken due to natural subtraction)
- Add D_one : D 1 = A 1
- Add D_two : D 2 = !![2, 0; 0, 2] (disconnected diagram)

JohnnyTeutonic commented:
> @JohnnyTeutonic This looks good except for the fix for `D 2`. Are you still planning on getting to that?

Ah yes, I've just committed that change. Fixed the D definition and added D_one, D_two.

ocfnash commented on Mathlib/Data/Matrix/Cartan.lean:
```diff
@@ -166,14 +168,40 @@ theorem D_off_diag_nonpos (i j : Fin n) (h : i ≠ j) : D n i j ≤ 0 := by
+                           -1,  2,  0;
+                           -1,  0,  2] := by decide
+
+example : (D 3).reindex c[0, 1] c[0, 1] = A 3 := by decide
```
```suggestion
theorem D_three' : (D 3).reindex c[0, 1] c[0, 1] = A 3 := by decide
```

ocfnash reviewed (approved):
Thanks!

ocfnash committed:
Update Mathlib/Data/Matrix/Cartan.lean

ocfnash commented:
bors merge
